Add fix orient/eco to USER-MISC

vmohles opened the pull request:
**Summary**

This is the Energy Conserving Orientational (ECO) Driving Force, which is an improved replacement/substitute for fix_orient. This is the efficient new version by Schratt and Mohles, as documented. It adds a potential energy to every atom depending on the two relevant crystal orientations, in order to drive grain boundaries. Works for FCC or BCC, and probably other crystals as well.  

**Related Issues**

First implementation, as published.

**Author(s)**

Adrian A. Schratt and Volker Mohles,
Interdisciplinary Centre for Advanced Materials Simulation (ICAMS)
Ruhr-Universität Bochum
Universitätsstr. 150
44801 Bochum, Germany
volker.mohles@rub.de

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

new addition, should not break anything if unused, meant to replace/substitute fix_orient

**Implementation Notes**

new addition, tested and used with the example provided at various temperatures and driving force magnitudes. nothing else SHOULD be affected.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

The description of the method and application for the example are given here:
https://doi.org/10.1016/j.commatsci.2020.109774

vmohles commented:
I see no reasons/explanations in the "Details" on why build_docs failed, or I don't understand them. Can anybody help?
thanks, Volker Mohles

rbberger commented:
This seems to be caused by a spellcheck failure in build_docs. See https://ci.lammps.org/job/dev/job/pull_requests/job/build_docs/251/warnings2Result/

You can check this on your workstation by running `make spelling` in the `doc` folder.

If a word is a false positive, please add it to [doc/utils/sphinx-config/false_positives.txt](https://github.com/lammps/lammps/blob/master/doc/utils/sphinx-config/false_positives.txt)

akohlmey commented on src/MISC/fix_eco_force.h:
```diff
@@ -0,0 +1,91 @@
+    double norm_fac;            // normalization constant
+    double inv_norm_fac;        // inverse normalization constant
+
+    Nbr *nbr;                   // pointer on array of Nbr structs
```
please put just a forward declaration here and define this struct in the .cpp file only.

akohlmey commented on src/MISC/fix_eco_force.h:
```diff
@@ -0,0 +1,91 @@
+
+namespace LAMMPS_NS {
+
+#define FIX_ECO_FORCE_MAX_NEIGHBORS 24
```
this define can move with the full declaration of the struct into the .cpp file.

akohlmey commented on src/MISC/fix_eco_force.h:
```diff
@@ -0,0 +1,91 @@
+
+#ifdef FIX_CLASS
+
+FixStyle(eco/force,FixECOForce)
```
if this is so similar to fix orient/XXX, it should be named similarly: how about `fix orient/eco` or `fix orient/ecodf`?
this will make it *much* easier for people to find it and try it out. Also, we are always looking for descriptive names of commands (at least for new contributions) and `eco/force` has many associations but nothing similar to what it does.
consequently with the renaming of the fix style name, also the source files and documentation and examples should be renamed to be consistent with the name.

akohlmey commented on examples/eco_force/sigma5.ori:
```diff
@@ -0,0 +1,6 @@
+2.228913 -1.114456 0.000000
+0.557228 1.671685 1.76212
+1.671685 -0.557228 1.76212
+2.228913 1.114456 0.000000
```
please make sure that *all* files have a LF at the end of the file. files without result in short reads on some platforms.

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+   Contributing authors: Adrian A. Schratt and Volker Mohles (ICAMS)
+------------------------------------------------------------------------- */
+
+#include "math.h"
```
please change the include statements to follow our conventions as described in `doc/include-file-conventions.md`

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+  peratom_freq = 1;         //
+
+  // parse input parameters
+  u_0 = atof(arg[3]);
```
please use `utils::numeric()` (or `force->numeric()`) instead of `atof()`

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+    char *result;
+    int count;
+
+    FILE *infile = fopen(dir_filename, "r");
```
please use `force->open_potential()` to open such files, so they can be placed into a common folder and found by setting the environment variable `LAMMPS_POTENTIALS`.

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+
+    FILE *infile = fopen(dir_filename, "r");
+    if (infile == NULL) error->one(FLERR, "Fix eco/force file open failed");
+    for (int i = 0; i < 6; ++i) {
```
we very recently added new classes to simplify reading potential and parameter files (see PR #2113). There is an update to it in PR #2122 . Since that is likely to be merged before this PR, you may want to look into adopting that code for reading in data from a file.  We are working on replacing ad hoc file readers wherever possible with reasonable effort throughout all of LAMMPS, so you could save us some time and effort and make the reading more robust.

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+  MPI_Bcast(&inv_eta, 1, MPI_DOUBLE, 0, world);                         // communicate inverse threshold
+
+  // set comm size needed by this Fix
+  if (u_0 != 0.0) comm_forward = 110;
```
how are these numbers determined? please add a comment so that if somebody wants to update/refactor the code it is possible to adjust them, if needed.

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+  if (me == 0) {
+    int neigh;
+    get_norm(neigh);
+    printf("fix eco/force: cutoff=%f norm_fac=%f neighbors=%i\n", r_cut, norm_fac, neigh);
```
please never use printf() like this. instead output to either the `screen` or both `screen` and `logfile` FILE pointers, if non-NULL, as it is done throughout the code. Those can be configured to be turned off (and thus any output unlike printf()s).
since the last patch release yesterday, there is also the option to use `utils::logmesg()` to output such messages
and `fmt::format()` or `fmt::printf()` to have a simpler code (please see https://fmt.dev for how to use this feature, or other uses in the current master branch).

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+
+  // check parameters
+  if (r_cut > force->pair->cutforce) {
+    error->all(FLERR, "Pair cutoff too small: use 'hybrid/overlay' to set the pair cutoff larger than the fix eco/force cutoff");
```
why do you need to have an increased force cutoff? and why through this unusual method?

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+      // if k is a ghost atom (already comm'd), its IDs are already global
+
+      id = nbr[k].id[j];
+      if (k < nlocal) id = tag[id];
```
you need to use the `ubuf()` union here and all similar cases, or this will not work for large systems

akohlmey commented on src/MISC/fix_eco_force.cpp:
```diff
@@ -0,0 +1,642 @@
+      nbr[i].delta[j][0] = buf[m++];
+      nbr[i].delta[j][1] = buf[m++];
+      nbr[i].delta[j][2] = buf[m++];
+      nbr[i].id[j] = static_cast<tagint> (buf[m++]);
```
see above. all integers that could be 64-bit have to converted back from integer to double and back using the `ubuf` union.

akohlmey commented on examples/eco_force/Ni.eam.fs:
```diff
@@ -0,0 +1,6006 @@
+Sourse: The potential parameters were provided by Prof. G.J. Ackland (g.j.ackland@ed.ac.uk). Please, contact him if you have any questions.
```
please do not add copies of potential files, if it can be avoided. rather use a (relative) symbolic link to a potential already present in the potentials folder. similar to other examples.

akohlmey commented on doc/src/fix_eco_force.rst:
```diff
@@ -0,0 +1,167 @@
+Restrictions
+""""""""""""
+
+This fix is part of the MISC package. It is only enabled if LAMMPS was
```
as a user contributed style, this should be added to the USER-MISC package instead of MISC. consequently, also the example files should be moved to `examples/USER/misc/<style name>`

akohlmey commented on doc/src/fix_eco_force.rst:
```diff
@@ -0,0 +1,167 @@
+..  index:: fix eco/force
```
this is missing full integration into the lists of commands in the manual. please see the output of `make style_check` in the doc folder.

akohlmey commented:
@vmohles thanks for the updates. i will check it out later.

in the future, please add your comments and responses to the github website for the pull request, not the commit message, as the website is where we manage and coordinate the review process. You can also reply to individual comments or discuss suggestions or suggest alternatives there.
the purpose to do this is to have communications about development associated with the changes so one can review discussions and decisions later in case of problems or suggestions for changes.

a compiler with working/enabled c++11 support is a *requirement* for LAMMPS now. we are in the process of some significant refactoring to take advantage of that and also to get rid of some ugly hacks because of trying hard to write LAMMPS in "C with classes" in the past.
because of the refactoring, we may delay merging feature additions for a few weeks unless they are up-to-date with the current master (so that we minimize merge conflicts on our side).

whether pull requests are (still) open can be seen from: https://github.com/lammps/lammps/pulls
I hope that I can merge several refactoring pull requests later today after a discussion with other LAMMPS developers.

The shortest way to update a feature branch from a fork to the current LAMMPS master would be:
```
git checkout <my feature branch>
git fetch https://github.com/lammps/lammps.git master
git merge FETCH_HEAD
<resolve any possible merge conflicts>
git push origin
```

akohlmey commented:
@vmohles i've updated your branch to the latest master, did some cosmetic cleaning (no trailing whitespace, commented out unused arguments), and updated the tests to align better with our conventions.

i also changed the code that was mixing local atom indices and global indices to just store only local indices and always communicate the global index. however, i am not convinced that that a forward communication is what you need to do to get neighbors of neighbors. since LAMMPS is not applying minimum image conventions, but has explicit copies you not only want the local index, but the one that is the closest neighbor. there is a method for that in the Domain class.

but in general, if you need to store neighbors of neighbors, you should simply instruct the neighborlist code to also store neighbors for ghost atoms. and then when building your short neighbor list, add a check for each neighbor that is a ghost atom whether it has its neighbors stored in your short neighbor list or else update that entry as well from the neighbors of the ghost atoms.

I have altered the neighbor list request accordingly, but have not implemented the modified short neighborlist build. that is better for your to decide what to do.

the reason i have been looking into this is that there is a significant difference in the log file when running with 1 MPI rank and 4 MPI ranks. i made some minimal test with replacing fix orient/fcc and that doesn't show that behavior. so this indicates something that is not yet 100% correct.. and the unusual way of communicating neighbors of neighbors was a first candidate to look at.

vmohles commented on src/USER-MISC/fix_orient_eco.cpp:
```diff
@@ -203,6 +201,7 @@ void FixOrientECO::init() {
   neighbor->requests[irequest]->fix = 1;
   neighbor->requests[irequest]->half = 0;
   neighbor->requests[irequest]->full = 1;
+  neighbor->requests[irequest]->ghost = 1;
```
It is not necessary to access neighbors of neighbors. the local neighbor list was created and used only because it is significantly shorter than the neighbor list used for the MD potentials. Therefore, this line should be removed! We can't do that here ourselves, it seems.

akohlmey replied:
@vmohles you are currently accessing neighbors of neighbors indirectly through the Nbr struct. Otherwise you would not need to do a forward communication to send it to ghost atoms. However, this is not always listing the correct neighbors, since you may have multiple copies of the same "tag" as ghost atoms and you would need to know the index of the atom closest to center atom. 

That means, if you include neighbors of ghost atoms in the neighbor list directly, you could construct the list of struct Nbr elements directly locally without the forward communication. For all atoms "j" that are confirmed neighbors of the short neighbor list, you would also look up and store the neighbors within the cutoff into the corresponding data structure (unless they are already included). You may be getting away with your approach in most cases, since you also store the distance vector to the center atom.

this is **your** branch, and of course you can change it any way your like. since you allowed maintainers to push their changes into your branch, all you have to do is to pull the changes from your github fork into your local fork first, then edit/apply your modifications and push it back.

but since i noticed another change that needs to be made i am apply this now and will push it (**never** commit changes to the makefiles, if you need a customized makefile, make a copy to src/MAKE/MINE and modify that copy and don't add/commit that. those copies will override the version in the other folders).

akohlmey commented:
@vmohles to summarize, i think i have adjusted everything i can do myself to resolve all remaining issues except for one item: the difference between serial and parallel execution.
this indicates, that something is off somewhere, but that is something for you to sort out. it would take me far too much time to first dig into what it is exactly that you are doing to be able to find out why things are not consistent and then come up with a solution.

vmohles commented:
We have reworked the code such that our reduced neighbor list is no longer needed.
This reduced the ghost atom communication significantly, and so we found the error
that you detected. THANKS! 
The MPI ranking conflict is now removed, 
and the simplification also made the code quite a bit faster and more memory efficient, 
both parallel and serial.

Now I have the trouble that I can’t commit to the new code or push my repository, because 
it seems someone else has altered my branch (fix_orient_eco.cpp) on Github. 
But I can’t pull my branch either!

GitHub is chaos to me. How can I proceed to update our files to our new, conflict free version?

Thanks,
Volker & Adrian


------------------------------------------------------
PD Dr. rer. nat. Volker Mohles
Interdisciplinary Centre for Advanced Materials Simulation
Ruhr-Universität Bochum
Universitätsstr. 150
44801 Bochum, Germany
Tel. +49 234 32 22295

> Am 08.06.2020 um 15:44 schrieb Axel Kohlmeyer <notifications@github.com>:
> 
> 
> @vmohles to summarize, i think i have adjusted everything i can do myself to resolve all remaining issues except for one item: the difference between serial and parallel execution.
> this indicates, that something is off somewhere, but that is something for you to sort out. it would take me far too much time to first dig into what it is exactly that you are doing to be able to find out why things are not consistent and then come up with a solution.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

akohlmey commented:
> We have reworked the code such that our reduced neighbor list is no longer needed. This reduced the ghost atom communication significantly, and so we found the error that you detected. THANKS! The MPI ranking conflict is now removed, and the simplification also made the code quite a bit faster and more memory efficient, both parallel and serial. Now I have the trouble that I can’t commit to the new code or push my repository, because it seems someone else has altered my branch (fix_orient_eco.cpp) on Github. But I can’t pull my branch either! GitHub is chaos to me. How can I proceed to update our files to our new, conflict free version? Thanks, Volker & Adrian

Thanks for the update, this is great news.

No worries about your git troubles. There is a bit of a learning curve. We understand that, but the benefits for the LAMMPS developers are very large compared to the alternative, that we don't want to go back. In the interest of expediting the process, please send the updated files to my personal e-mail and I will see to it that they are included and the pull request updated. We can discuss at a later point, e.g. when some minor updates would be needed, how to sort this all out. It is likely not that difficult, just unfamiliar. We are also about to make a public announcement for a LAMMPS development slack channel, which might be an easier path to help people with processing and updating pull requests through git, since it can be done interactively.

akohlmey commented:
@sjplimp this now checks out, is valgrind clean. has no more neighborlist complications, and produces consistent results in serial and parallel and thus could be merged for today's(?) patch release.
